### PR TITLE
elpa: add 2017.05.002 and 2017.05.003

### PR DESCRIPTION
**Pull request by davydden**
tested on Ubuntu16+gcc5.4.0.

### CODE REVIEW

File: var/spack/repos/builtin/packages/elpa/package.py [hunk 1]
```diff
@@ -76,7 +75,14 @@ def setup_environment(self, spack_env, run_env):
         spack_env.set('SCALAPACK_LDFLAGS', spec['scalapack'].libs.joined())
```
Comment (davydden) [inline var/spack/repos/builtin/packages/elpa/package.py]: btw, @alalazo why did you move those from being options to `configure` here?

File: var/spack/repos/builtin/packages/elpa/package.py [hunk 2]
```diff
@@ -76,7 +75,14 @@ def setup_environment(self, spack_env, run_env):
...
 
     def configure_args(self):
-        options = []
+        # TODO: set optimum flags for platform+compiler combo, see
+        # https://github.com/hfp/xconfigure/tree/master/elpa
+        # also see:
+        # https://src.fedoraproject.org/cgit/rpms/elpa.git/
+        options = [
+            'FCFLAGS=-O3 -march=native -ffree-line-length-none',
+            'CFLAGS=-O3 -march=native'
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/elpa/package.py]: I am ccing @lee218llnl as I remember we had a discussion (was it `cp2k`?) where flags like `-march=<something>` were considered controversial. Also isn't `-march` GCC specific?

Comment (davydden) [reply var/spack/repos/builtin/packages/elpa/package.py]: for the record, without these flags configure fails testing `sse` related assembly. I can wrap it into `if spec.satisfies('%gcc')`.

Comment (davydden) [reply var/spack/repos/builtin/packages/elpa/package.py]: > Also isn't -march GCC specific?

i don't think so, it certainly works with `clang` and `intel`.

Comment (alalazo) [reply var/spack/repos/builtin/packages/elpa/package.py]: > i don't think so, it certainly works with clang and intel.

Thanks, didn't have `intel` at hand to double check.

Comment (davydden) [reply var/spack/repos/builtin/packages/elpa/package.py]: > didn't have intel at hand to double check.

i also don't :smile: but I know for sure that my colleagues use this flag with `intel`.

Comment (lee218llnl) [reply var/spack/repos/builtin/packages/elpa/package.py]: The Intel compiler does accept the -march option. The controversy with -march=native is for sites with a shared file system that is mounted on different architectures. For example, we have a system that supports AVX2, so compiling with the Intel compiler and -march=native enables AVX2 instructions. If I then run on older hardware that doesn't have AVX2, I see this:

Please verify that both the operating system and the processor support Intel(R) MOVBE, F16C, FMA, BMI, LZCNT and AVX2 instructions.

Comment (davydden) [reply var/spack/repos/builtin/packages/elpa/package.py]: @lee218llnl anything in spack like `if sepc.satisfies(platform ==)` i should add here in order to avoid the issue you described?

Comment (lee218llnl) [reply var/spack/repos/builtin/packages/elpa/package.py]: I don't think the platform or anything else currently in the arch specification will help here. You could try using the `-mtune=generic` flag instead of `-march=native`, but you wouldn't get optimal performance. @becker33 do you have any suggestions for how to deal with vectorization flags?

Comment (becker33) [reply var/spack/repos/builtin/packages/elpa/package.py]: There isn't anything currently in the arch specification. I am working on this in #3206, but there's still some work to do there. That PR will make the target be as specific as the gcc arguments to `-march`.

Comment (davydden) [reply var/spack/repos/builtin/packages/elpa/package.py]: I think I will make a variant which is true by default. This shall work for majority of cases

**Comment by davydden** [comment]
@alalazo this is ready for another review.

**Comment by davydden** [comment]
@alalazo ping.

File: var/spack/repos/builtin/packages/elpa/package.py (hunk 1 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/elpa/package.py]: To group together all the changes to the environment variables in the function that is meant to change the environment variables for this package.

**Comment by davydden** [comment]
@alalazo ping.

File: var/spack/repos/builtin/packages/elpa/package.py [hunk 3]
```diff
@@ -76,7 +73,17 @@ def setup_environment(self, spack_env, run_env):
         spack_env.set('SCALAPACK_LDFLAGS', spec['scalapack'].libs.joined())
```
Comment (baip) [inline var/spack/repos/builtin/packages/elpa/package.py]: I'm just getting familiar with Spack, and trying to debug elpa on my systems (#6143, #6144). I'm wondering if the use of `setup_environment()` is appropriate in this package. According to #2016: the purpose of `setup_environment()` seems to be setting up environment modules, not setting up environment for `configure()` or `build()`.

Comment (davydden) [reply var/spack/repos/builtin/packages/elpa/package.py]: See above, that's what the reasoning of @alalazo who move it from configure to `setup_environment()`

> To group together all the changes to the environment variables in the function that is meant to change the environment variables for this package.

Comment (davydden) [reply var/spack/repos/builtin/packages/elpa/package.py]: > According to #2016: the purpose of setup_environment() seems to be setting up environment modules

it's called twice, https://github.com/spack/spack/issues/2016#issuecomment-258936647 so it's not only for modules

Comment (baip) [reply var/spack/repos/builtin/packages/elpa/package.py]: Thanks for the clarification. If those environment variables are not needed in modulefiles, can they be moved to `configure_args()`? Could you please take a look at my tentative changes [here](https://github.com/baip/spack/commit/af4ed0b0c700c6444fbab022cd1adf501c04fb19)? With these changes, my issues with GCC 6.4.0/7.2.0 (#6143) or Intel 2018.0 (#6144) went away. Before, in the second call to `setup_environment()`, `spec['mpi'].mpicc` somehow became undefined.

Comment (davydden) [reply var/spack/repos/builtin/packages/elpa/package.py]: Please open a PR with suggested changes so that everyone can comment

**Comment by davydden** [comment]
@alalazo do you have further comments to `elpa` PR? It's more than 2 months already, i would like it to have merged. I kept old version as preferred and tested `2016.05.044`, `2017.05.002` and `2017.05.003` on Ubuntu 16 with gcc 5.4.0. As a reminder I added a comment why compiler flags are needed in this case.

**Comment by alalazo** [comment]
I am fine with it, given the comments. If @lee218llnl or other people maintaining installations are also fine with `march=native` there, I think we can merge this without issues. Waiting a few days for their confirmation (or lack thereof) before merging.

**Comment by alalazo** [comment]
@davydden Please ping me again in a couple of days if nothing happens here. With almost 650 issues and 200 PRs it can be easy to forget things :smile:

**Comment by davydden** [comment]
Thanks for prompt reply @alalazo .

> Please ping me again in a couple of days if nothing happens here.

will do.

> If @lee218llnl or other people maintaining installations are also fine with march=native there

for the record, one can disable those flags with `~optflags` and set up custom tailored flags.

**Comment by lee218llnl** [comment]
I don't install elpa so don't let me hold you up on getting this merged.

**Comment by davydden** [comment]
@alalazo ping-ping.

**Comment by alalazo** [comment]
I merged the PR. Worst case (very unlikely) we'll revert it and split it into 2 parts.
